Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+HTTP2Connections.swift

```diff
@@ -0,0 +1,532 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2021 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIOCore
+
+extension HTTPConnectionPool {
+    private struct HTTP2ConnectionState {
+        private enum State {
+            /// the pool is establishing a connection. Valid transitions are to: .idle, .available and .closed
+            case starting
+            /// the connection is waiting to retry to establish a connection. Transition to .closed. From .closed
+            /// a new connection state must be created for a retry.
+            case backingOff
+            /// the connection is active and is able to run requests. Valid transitions to: .draining and .closed
+            case active(Connection, maxStreams: Int, usedStreams: Int, lastIdle: NIODeadline)
+            /// the connection is active and is running requests. No new requests must be scheduled.
+            /// Valid transitions to: .draining and .closed
+            case draining(Connection, maxStreams: Int, usedStreams: Int)
+            /// the connection is closed
+            case closed
+        }
+
+        var isActive: Bool {
+            switch self.state {
+            case .starting, .backingOff, .draining, .closed:
+                return false
+            case .active:
+                return true
+            }
+        }
+
+        var isStarting: Bool {
+            switch self.state {
+            case .starting, .backingOff:
+                return true
+            case .active, .draining, .closed:
+                return false
+            }
+        }
```

Comment:
It is not clearly communicated in my mind that starting is connecting _and_ backingOff.